dashboard/modules/runtime_env/runtime_env_agent.py

```diff
@@ -21,20 +22,28 @@ class RuntimeEnvAgent(dashboard_utils.DashboardAgentModule,
     def __init__(self, dashboard_agent):
         super().__init__(dashboard_agent)
         self._runtime_env_dir = dashboard_agent.runtime_env_dir
+        self._setup = import_attr(dashboard_agent.runtime_env_setup_hook)
         runtime_env.PKG_DIR = dashboard_agent.runtime_env_dir
 
     async def CreateRuntimeEnv(self, request, context):
+        logger.info("Creating runtime env: %s.",
+                    request.serialized_runtime_env)
         runtime_env_dict = json.loads(request.serialized_runtime_env or "{}")
         uris = runtime_env_dict.get("uris")
         if uris:
-            logger.info("Creating runtime env with uris %s", repr(uris))
             # TODO(guyang.sgy): Try `ensure_runtime_env_setup(uris)`
             # to download packages.
             # But we don't initailize internal kv in agent now.
             pass
-
+        result = self._setup(request.serialized_runtime_env,
+                             self._runtime_env_dir)
+        runtime_env_dict["result"] = result
```

Comment:
It looks like the runtime env dict is being used as a way to communicate from the agent setup function to the worker_setup function, probably because the existing implementation of the setup function was unfortunately not very pluggable.  It seems fine for now, but do you have any ideas on what the long term solution should be for this kind of communication? Or is this okay for the long term?  I'm just wondering if we need some kind of better API for this.

Reply:
Yep, I also have thought about this question. But I didn't find any better way to do this because  `setup_worker` just parses params from command line.

Reply:
I think we should separate the "runtime env spec" (which is specified by the user) and the "runtime env context (which contains the context info about a runtime env, such as path to a conda). 
Instead of inserting the context info back the the `RuntimeEnv` object, it's better to define a new class `RuntimeEnvContext` and use a new field in the reply. 
This would help improve code readability and maintainability.